File: testapps/PHP/Plugins/AutoGen/app/CommonPlugin.php
<?php
#-------------------------------------------------------------------------------
# Copyright 2019 NAVER Corp
# 
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#-------------------------------------------------------------------------------

namespace Plugins\AutoGen\app;
use Pinpoint\Plugins\Common\PinTrace;

///@hook:app\common\AppDate::outputDate
///@hook:app\common\AopFunction::test_func1
///@hook:app\common\AopFunction::test_func2  app\util\Foo::test_func_2
///@hook:app\common\Person::eat app\common\Teacher::eat app\common\Student::eat app\common\Doctor::other app\common\Doctor::eat
///@hook:app\common\AbstractStudent::eat app\common\AbstractStudent::drink app\common\AbstractStudent::breath
///@hook:app\common\AbstractStudent::eat app\common\AbstractStudent::drink app\common\AbstractPerson::breath
///@hook:app\common\InterfaceStudent::setVariable app\common\InterfaceStudent::getHtml app\common\InterfaceStudent::other
///@hook:app\common\TestClone::setColor app\common\TestClone::getColor app\common\TestClone::__clone
///@hook:app\common\TestRecursion::test_cumsum1 app\common\TestRecursion::test_cumsum2
///@hook:app\common\TestAnonymous::createbell
///@hook:app\common\TestStatic::static_func
///@hook:app\common\TestFinalClass::test app\common\TestFinalFunc::test
///@hook:app\common\FuncInTrait::traitfunc app\common\TestTrait::test
///@hook:app\common\CallBackFunc::fnCallback1 app\common\CallBackFunc::fnCallback2
///@hook:app\common\LevelClass::public_function app\common\LevelClass::protected_function app\common\LevelClass::private_function
///@hook:app\common\OverRideLevel::public_function app\common\OverRideLevel::protected_function app\common\OverRideLevel::private_function
///@hook:app\common\TestArgs::test_args
///@hook:app\common\TestReturn::test_return
///@hook:app\common\TestError::throwException app\common\TestError::test_cumsum_e1 app\common\TestError::caughtException app\common\TestError::error app\common\TestError::testParseError app\common\TestError::testTypeError app\common\TestError::testArithmeticError app\common\TestError::testDivisionByZeroError app\common\TestError::testAssertionError
class CommonPlugin extends PinTrace
{
    ///@hook:app\common\DBcontrol::connectDb
    public function onBefore(){
        pinpoint_add_clue("stp",PP_PHP_METHOD);
        pinpoint_add_clues(PP_PHP_ARGS,print_r($this->args,true));
    }

    ///@hook:app\common\DBcontrol::getData1 app\common\DBcontrol::\array_push
    public function onEnd(&$ret){
        pinpoint_add_clues(PP_PHP_RETURN,print_r($ret,true));
    }

    ///@hook:app\common\DBcontrol::getData2
    public function onException($e){
        pinpoint_add_clue("EXP",$e->getMessage());
    }
}

File: plugins/PY/pinpointPy/CommonPlugin.py
# ------------------------------------------------------------------------------
#  Copyright  2020. NAVER Corp.                                                -
#                                                                              -
#  Licensed under the Apache License, Version 2.0 (the "License");             -
#  you may not use this file except in compliance with the License.            -
#  You may obtain a copy of the License at                                     -
#                                                                              -
#   http://www.apache.org/licenses/LICENSE-2.0                                 -
#                                                                              -
#  Unless required by applicable law or agreed to in writing, software         -
#  distributed under the License is distributed on an "AS IS" BASIS,           -
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    -
#  See the License for the specific language governing permissions and         -
#  limitations under the License.                                              -
# ------------------------------------------------------------------------------


from pinpointPy import Common, Defines, pinpoint
import random


class PinpointCommonPlugin(Common.PinTrace):

    def onBefore(self, parentId, *args, **kwargs):
        trace_id, args, kwargs = super().onBefore(parentId, *args, **kwargs)
        pinpoint.add_trace_header(
            Defines.PP_INTERCEPTOR_NAME, self.getUniqueName(), trace_id)
        pinpoint.add_trace_header(
            Defines.PP_SERVER_TYPE, Defines.PP_METHOD_CALL, trace_id)
        return trace_id, args, kwargs

    def onEnd(self, traceId, ret):
        # pinpoint.add_trace_header_v2(Defines.PP_RETURN, str(ret), traceId)
        super().onEnd(traceId, ret)
        return ret

    def onException(self, traceId, e):
        pinpoint.add_exception(str(e), traceId)
        raise e


class AsyncCommonPlugin(Common.AsyncPinTrace):

    # -> tuple[int, Any, dict[str, Any]]:
    def onBefore(self, parentId, *args, **kwargs):
        traceId, args, kwargs = super().onBefore(parentId, *args, **kwargs)
        pinpoint.add_trace_header(
            Defines.PP_INTERCEPTOR_NAME, self.getUniqueName(), traceId)
        pinpoint.add_trace_header(
            Defines.PP_SERVER_TYPE, Defines.PP_METHOD_CALL, traceId)
        return traceId, args, kwargs

    def onEnd(self, traceId, ret):
        super().onEnd(traceId, ret)

    def onException(self, traceId, e):
        pinpoint.add_exception(str(e), traceId)


class HookTargetPlugins(Common.PinTrace):
    def onBefore(self, parentId: int,  *args, **kwargs):

        traceId, args, kwargs = super().onBefore(parentId, *args, **kwargs)
        pinpoint.add_trace_header(
            Defines.PP_INTERCEPTOR_NAME, self.getUniqueName(), traceId)
        pinpoint.add_trace_header(
            Defines.PP_SERVER_TYPE, Defines.P_INVOCATION_CALL_TYPE, traceId)

        async_id = random.randint(0, 9999)
        pinpoint.add_trace_header(
            Defines.PP_ASYNC_CALL_ID, f'{async_id}', traceId)

        sequence_id = pinpoint.get_sequence_id(traceId)

        tid = pinpoint.get_context(Defines.PP_TRANSACTION_ID, traceId)
        seq_id = pinpoint.get_context(Defines.PP_SPAN_ID, traceId)
        app_name = pinpoint.get_context(Defines.PP_APP_NAME, traceId)
        app_id = pinpoint.get_context(Defines.PP_APP_ID, traceId)

        if 'target' in kwargs:
            origin_target = kwargs['target']

            def pp_new_entry_func(*args, **kwargs):
                # start trace
                thread_trace_id = pinpoint.with_trace(0)
                self.setCurrentTraceNodeId(thread_trace_id)
                pinpoint.add_trace_header(
                    Defines.PP_APP_NAME, app_name, thread_trace_id)
                pinpoint.add_context(
                    Defines.PP_APP_NAME, app_name, thread_trace_id)

                pinpoint.add_trace_header(
                    Defines.PP_APP_ID, app_id, thread_trace_id)
                pinpoint.add_context(
                    Defines.PP_APP_ID, app_id, thread_trace_id)

                pinpoint.add_trace_header(
                    Defines.PP_SPAN_ID, seq_id, thread_trace_id)
                pinpoint.add_context(
                    Defines.PP_SPAN_ID, seq_id, thread_trace_id)

                pinpoint.add_trace_header(
                    Defines.PP_TRANSACTION_ID, tid, thread_trace_id)
                pinpoint.add_context(
                    Defines.PP_TRANSACTION_ID, tid, thread_trace_id)

                pinpoint.add_trace_header(
                    Defines.PP_SERVER_TYPE, Defines.PYTHON, thread_trace_id)
                pinpoint.set_async_context(
                    thread_trace_id, async_id, sequence_id)

                if callable(origin_target):
                    @PinpointCommonPlugin(origin_target.__name__)
                    def call_origin_target(*args, **kwargs):
                        origin_target(*args, **kwargs)

                    call_origin_target(*args, **kwargs)

                pinpoint.end_trace(thread_trace_id)

            kwargs['target'] = pp_new_entry_func

        return traceId, args, kwargs

    def onEnd(self, traceId, ret):
        super().onEnd(traceId, ret)
        return ret
